[MIPS] Fix missing ANDI optimization

yingopq opened the pull request:
1. Add MipsPat to optimize (andi (srl (truncate i64 $1), x), y) to (andi (truncate (dsrl i64 $1, x)), y).
2. Add MipsPat to optimize (ext (truncate i64 $1), x, y) to (truncate (dext i64 $1, x, y)).

The assembly result is the same as gcc.

Fix #42826

yingopq committed:
[MIPS] Fix missing ANDI optimization

1. Add MipsPat to optimize (andi (srl (truncate i64 $1), x), y) to
   (andi (truncate (dsrl i64 $1, x)), y).
2. Add MipsPat to optimize (ext (truncate i64 $1), x, y) to
   (truncate (dext i64 $1, x, y)).

The assembly result is the same as gcc.

Fix #42826

yingopq commented:
@topperc @wzssyqa Could you help review this patch, thanks!

brad0 commented:
@arichardson

topperc commented on llvm/lib/Target/Mips/Mips64InstrInfo.td:
```diff
@@ -830,6 +830,12 @@ def : MipsPat<(sra GPR64:$rt, (i32 (trunc GPR64:$rs))),
 def : MipsPat<(rotr GPR64:$rt, (i32 (trunc GPR64:$rs))),
               (DROTRV GPR64:$rt, (EXTRACT_SUBREG GPR64:$rs, sub_32))>,
               ISA_MIPS3, GPR_64;
+def : MipsPat<(and (srl (i32 (trunc GPR64:$src)), immZExt5:$imm5), immZExt5:$value),
```
Can the `and` immediate be larger than 5 bits? I'm not very familiar with Mips encodings.

yingopq replied:
> Can the `and` immediate be larger than 5 bits? I'm not very familiar with Mips encodings.

Sorry, it can be larger than 5, it should be `immZExt16`. I would submit again. Thanks!

arichardson reviewed (approved):
I don't work on MIPS anymore but this looks correct to me.

One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.

yingopq commented:
> I don't work on MIPS anymore but this looks correct to me.
> 
> One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.

Yes, this would be best. Thanks! 
Need I split it into two patches and resubmit again?

arichardson commented:
> > I don't work on MIPS anymore but this looks correct to me.
> > One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.
> 
> Yes, this would be best. Thanks! Need I split it into two patches and resubmit again?

You could commit the baseline test directly and then just update this pull request.

yingopq commented:
> > > I don't work on MIPS anymore but this looks correct to me.
> > > One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.
> > 
> > 
> > Yes, this would be best. Thanks! Need I split it into two patches and resubmit again?
> 
> You could commit the baseline test directly and then just update this pull request.

OK, I submitted test https://github.com/llvm/llvm-project/pull/102278. Please help review, thanks!

arichardson commented:
> > > > I don't work on MIPS anymore but this looks correct to me.
> > > > One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.
> > > 
> > > 
> > > Yes, this would be best. Thanks! Need I split it into two patches and resubmit again?
> > 
> > 
> > You could commit the baseline test directly and then just update this pull request.
> 
> OK, I submitted test #102278. Please help review, thanks!

I've merged that other commit, once this one has been rebased I can go ahead and merge it. Thanks!

yingopq committed:
Merge branch 'main' into Fix_bug_issue_42826

yingopq commented:
> > > > > I don't work on MIPS anymore but this looks correct to me.
> > > > > One nit: I think it would be good to pre-commit the test with the existing codegen first and then have this patch show the codegen improvement.
> > > > 
> > > > 
> > > > Yes, this would be best. Thanks! Need I split it into two patches and resubmit again?
> > > 
> > > 
> > > You could commit the baseline test directly and then just update this pull request.
> > 
> > 
> > OK, I submitted test #102278. Please help review, thanks!
> 
> I've merged that other commit, once this one has been rebased I can go ahead and merge it. Thanks!

Thanks, I have updated this pull request.
